### PR TITLE
Feature: Server-Sent Events

### PR DESCRIPTION
**Pull request by 0xi4o**
- Uses SSE for streaming responses instead of socket.io

Related: https://github.com/FlowiseAI/Flowise/pull/3125

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,11 +17,12 @@
   "license": "MIT",
   "dependencies": {
     "@babel/core": "^7.22.1",
+    "@microsoft/fetch-event-source": "^2.0.1",
     "@ts-stack/markdown": "^1.4.0",
     "device-detector-js": "^3.0.3",
     "lodash": "^4.17.21",
+    "lodash.clonedeep": "^4.5.0",
```
Comment (HenryHengZJ) [inline package.json:24]: do we need to specifically import this?

Comment (0xi4o) [reply package.json]: Removed it

**Comment by HenryHengZJ** [comment]
can we change the order? to have the feedback after source documents:
image

File: src/components/Bot.tsx [hunk 2]
```diff
@@ -240,11 +242,11 @@ export const Bot = (botProps: BotProps & { class?: string }) => {
     { equals: false },
   );
 
-  const [socketIOClientId, setSocketIOClientId] = createSignal('');
   const [isChatFlowAvailableToStream, setIsChatFlowAvailableToStream] = createSignal(false);
   const [chatId, setChatId] = createSignal(
     (props.chatflowConfig?.vars as any)?.customerId ? `${(props.chatflowConfig?.vars as any).customerId.toString()}+${uuidv4()}` : uuidv4(),
   );
+  const [isMessageStopping, setIsMessageStopping] = createSignal(false);
```
Comment (HenryHengZJ) [inline src/components/Bot.tsx:249]: we dont have abort functionality yet, so we can remove this for now as its not being used

File: src/components/Bot.tsx [hunk 3]
```diff
@@ -334,34 +336,21 @@ export const Bot = (botProps: BotProps & { class?: string }) => {
...
-    agentReasoning: IAgentReasoning[] = [],
-    action: IAction,
-    resultText: string,
-  ) => {
+  const updateLastMessage = (text: string) => {
     setMessages((data) => {
       const updated = data.map((item, i) => {
         if (i === data.length - 1) {
-          if (resultText && !hasSoundPlayed) {
+          if (!hasSoundPlayed) {
```
Comment (HenryHengZJ) [inline src/components/Bot.tsx:343]: this will play the notification sound for every event, we only want at the end

File: src/components/Bot.tsx [hunk 4]
```diff
@@ -334,34 +336,21 @@ export const Bot = (botProps: BotProps & { class?: string }) => {
...
-          return { ...item, message: item.message + text, messageId, sourceDocuments, fileAnnotations, agentReasoning, action };
+          return { ...item, message: item.message + text };
         }
         return item;
       });
       addChatMessage(updated);
       return [...updated];
     });
-
-    // Set hasSoundPlayed to false if resultText exists
```
Comment (HenryHengZJ) [inline src/components/Bot.tsx:361]: removing this will cause the sound not to play again, since `hasSoundPlayed` will always be true

Comment (HenryHengZJ) [reply src/components/Bot.tsx:361]: resolved by moving to onclose function

**Comment by HenryHengZJ** [comment]
> can we change the order? to have the feedback after source documents: image

resolved

**Comment by 0xi4o** [comment]
> pending changes to display error

@HenryHengZJ Can you give some context for this?

**Comment by HenryHengZJ** [comment]
Included as part of the fix:
- streaming of artifacts
- sanitize user input:
image

**Comment by HenryHengZJ** [comment]
> @HenryHengZJ Hello, is there way how to disable streming completly? I want to use proxy server and it would be more easy without streming.

not at the moment when using embedded, we can look into support disable streaming

**Pull request merged**